plat/kvm/x86/multiboot2.c

```diff
@@ -0,0 +1,180 @@
+
+#define multiboot2_crash(msg, rc)	ukplat_crash()
+
+/* Round up 'x' to the nearest multiple of 'MULTIBOOT_TAG_ALIGN' */
+#define ROUND_UP_TO_ALIGNMENT(x) \
+	(((x) + (MULTIBOOT_TAG_ALIGN - 1)) & ~(MULTIBOOT_TAG_ALIGN - 1))
+
+void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi);
+
+static inline void mrd_insert(struct ukplat_bootinfo *bi,
+			      const struct ukplat_memregion_desc *mrd)
+{
+	int rc;
+
+	if (unlikely(mrd->len == 0))
+		return;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, mrd);
+	if (unlikely(rc < 0))
+		multiboot2_crash("Cannot insert bootinfo memory region", rc);
+}
+
+/**
+ * Multiboot entry point called after lcpu initialization. We enter with the
+ * 1:1 boot page table set. Physical and virtual addresses thus match for all
+ * regions in the mapped range.
+ */
+void multiboot2_entry(struct lcpu *lcpu, unsigned long addr)
+{
+	struct ukplat_bootinfo *bi;
+	struct ukplat_memregion_desc mrd = {0};
+	multiboot_memory_map_t *m;
+	__u64 cmdline_str = {0};
+	__sz i, cmdline_len, mmap_entry_cnt;
+	__paddr_t start, end;
+	int rc;
+
+	bi = ukplat_bootinfo_get();
+	if (unlikely(!bi))
+		multiboot2_crash("Incompatible or corrupted bootinfo", -EINVAL);
+
+	/* We have to call this here as the very early do_uk_reloc32 relocator
+	 * does not also relocate the UKPLAT_MEMRT_KERNEL mrd's like its C
+	 * equivalent, do_uk_reloc, does.
+	 */
+	do_uk_reloc_kmrds(0, 0);
+
+	/* Ensure that the memory map contains the legacy high mem area */
+	rc = ukplat_memregion_list_insert_legacy_hi_mem(&bi->mrds);
+	if (unlikely(rc))
+		multiboot2_crash("Could not insert legacy memory region", rc);
+
+	/**
+	 * Initialize the first tag address, skipping total_size(4) and
+	 * reserved (4) fields of the boot information structure
+	 */
+	struct multiboot_tag *tag = (struct multiboot_tag *)(addr + 8);
+
+	/* Parse Multiboot2 tags */
+	for (; tag->type != MULTIBOOT_TAG_TYPE_END;
+	     tag = (struct multiboot_tag *)((multiboot_uint8_t *)tag
+					+ ROUND_UP_TO_ALIGNMENT(tag->size))) {
+		switch (tag->type) {
+		case MULTIBOOT_TAG_TYPE_CMDLINE:
+			struct multiboot_tag_string *tag_str;
+
+			tag_str = (struct multiboot_tag_string *)tag;
+			cmdline_str = (__u64)tag_str->string;
+			cmdline_len = tag_str->size;
+			break;
+		case MULTIBOOT_TAG_TYPE_MODULE:
+			struct multiboot_tag_module *mod_tag;
+
+			mod_tag = (struct multiboot_tag_module *)tag;
+
+			mrd.pbase = PAGE_ALIGN_DOWN(mod_tag->mod_start);
+			mrd.vbase = mrd.pbase;
+			mrd.pg_off = mod_tag->mod_start - mrd.pbase;
+			mrd.len = mod_tag->mod_end - mod_tag->mod_start;
+			mrd.pg_count = PAGE_COUNT(mrd.pg_off + mrd.len);
+			mrd.type = UKPLAT_MEMRT_INITRD;
+			mrd.flags = UKPLAT_MEMRF_READ;
+
+			#ifdef CONFIG_UKPLAT_MEMRNAME
+				strncpy(mrd.name,
+					(char *)(__uptr)mod_tag->cmdline,
+					sizeof(mrd.name) - 1);
+			#endif /* CONFIG_UKPLAT_MEMRNAME */
+
+			mrd_insert(bi, &mrd);
+			break;
+		case MULTIBOOT_TAG_TYPE_MMAP:
+			struct multiboot_tag_mmap *mmap_tag;
```

Comment:
No late declarations, do all declarations at the beginning. And this switch case is kind of fat, I would move it to its own function. This way you get to keep your declarations :P